File: src/System/Mesos/Raw/FrameworkId.hs
module System.Mesos.Raw.FrameworkId where
import           System.Mesos.Internal

type FrameworkIDPtr = Ptr FrameworkID

foreign import ccall unsafe "ext/types.h toFrameworkID" c_toFrameworkID :: ToID FrameworkIDPtr

foreign import ccall unsafe "ext/types.h fromFrameworkID" c_fromFrameworkID :: FromID FrameworkIDPtr

foreign import ccall unsafe "ext/types.h destroyFrameworkID" c_destroyFrameworkID :: FrameworkIDPtr -> IO ()

instance CPPValue FrameworkID where
  marshal x = do
    (strp, l) <- cstring $ fromFrameworkID x
    liftIO $ c_toFrameworkID strp (fromIntegral l)

  unmarshal p = fmap FrameworkID $ do
    ptrPtr <- alloc
    len <- liftIO $ c_fromFrameworkID p ptrPtr
    peekCString' (ptrPtr, len)

  destroy = c_destroyFrameworkID

File: ext/types/framework_id.cpp
#include <iostream>
#include "types.h"

using namespace mesos;

FrameworkIDPtr toFrameworkID(char* bs, int len)
{
  FrameworkIDPtr val = new FrameworkID();
  val->set_value(bs, len);
  return val;
}

int fromFrameworkID(FrameworkIDPtr p, char** poke)
{
  *poke = (char*) p->mutable_value()->data();
  return p->mutable_value()->size();
}

void destroyFrameworkID(FrameworkIDPtr p)
{
  delete p;
}
